File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        p1,p2 = [ 0 for _ in range(n)], [0 for _ in range(n)]
        p1[0] = nums[0]
        p2[-1] = nums[-1]
        
        for i in range(1,n):
            p1[i] = p1[i-1]*nums[i]
        for i in range(n-2,-1,-1):
            p2[i] = p2[i+1]*nums[i]
        output = [0 for _ in range(n)]
        output[n-1] = p1[n-2]
        output[0] = p2[1]
        for i in range(1,n-1):
            output[i] = p1[i-1]*p2[i+1]
        return output
        

File: Data Structures & Algorithms/permutations/submission-0.cpp
class Solution {
public:
vector<vector<int>> res;int n;
    void dfs(vector<int>& nums, vector<bool>& visited, int k, vector<int>& temp){
        if(k == n) res.push_back(temp);
        else{
            for(int i=0;i<n;++i){
                if(!visited[i]){
                    visited[i]= true;
                    temp.push_back(nums[i]);
                    dfs(nums,visited,k+1, temp);
                    temp.pop_back();
                    visited[i]=false;
                }
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        n = nums.size();
        vector<bool> visited(n, false);
        vector<int> temp;
        dfs(nums,visited,0,temp);
        return res;

    }
};
